src/sagemaker/local_session.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import datetime
+import logging
+import time
+
+import urllib3
+from botocore.exceptions import ClientError
+
+from sagemaker.image import SageMakerContainer
+from sagemaker.session import Session
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.WARNING)
+
+
+class LocalSagemakerClient(object):
+    """A SageMakerClient that implements the API calls locally.
+
+    Used for doing local training and hosting local endpoints. It still needs access to
+    a boto client to interact with S3 but it won't perform any SageMaker call.
+
+    Implements the methods with the same signature as the boto SageMakerClient.
+    """
+    def __init__(self, sagemaker_session=None):
+        """Initialize a LocalSageMakerClient.
+
+        Args:
+            sagemaker_session (sagemaker.session.Session): a session to use to read configurations
+                from, and use its boto client.
+        """
+        self.train_container = None
+        self.serve_container = None
+        self.sagemaker_session = sagemaker_session or LocalSession()
+        self.s3_model_artifacts = None
+        self.model_name = None
+        self.primary_container = None
+        self.role_arn = None
+        self.created_endpoint = False
+
+    def create_training_job(self, TrainingJobName, AlgorithmSpecification, RoleArn, InputDataConfig, OutputDataConfig,
```

Comment:
We need to decide if we want to support multiple training jobs or not.

If we expect customers to only do local training via estimators, then we don't. If we expect customers to use a LocalSession to do training, then we do. Create an issue for this, should frame the ambiguity.

The same applies to the hosting methods as well, for the different hosting entities - endpoint, model, endpoint config. Create an issue for this, again framing ambiguity.

Even if we only support one endpoint per session (as it is now), we need to be able to run multiple endpoints in different sessions concurrently. This means we need to flexible with ports in a way that we aren't now (I think). Create an issue for this.

Reply:
You are right, the port as of now is configurable but its not flexible at runtime. This is one of the things I considered improving as it should be fairly transparent to the users